### ISSUE TITLE
Text-field: Setting required with Component causes field to immediately be invalid. 

### ISSUE THREAD

**Issue by williamernest** [material-components/material-components-web#2029]
## Bugs

Follow the template below to ensure the quickest and most accurate response to your issue.

### What MDC Web Version are you using?

> 0.28.0

### What browser(s) is this bug affecting?

> Version 63.0.3239.132 (Official Build) (64-bit)

### What OS are you using?

> Debian 4.9.65-3+deb9u2 (2018-01-04) x86_64 GNU/Linux
> MacOS 10.13.2

### What are the steps to reproduce the bug?

1. Set one component to "required" using the JS component. 
2. Set another to required by adding "required" to the html element. 

### What is the expected behavior?

Setting a text-field as "required" should not immediately render the field as invalid until after the blur event. 

### What is the actual behavior?

The element set as a required element using the JS component has the --invalid class applied immediately. 

### Any other information you believe would be useful?
The trailing icon elements in the gif are using the Component to set the field as required. The leading icon elements are set to required using the html attribute 'required'. 
![final](https://user-images.githubusercontent.com/34519388/35235986-96f41472-ff5a-11e7-950d-dcae5b4cbe0c.gif)

**Comment by theenoahmason** [comment]
Hey I was just looking at this as I remembered this had been fixed a couple versions ago... As far as I can tell this is invalid, the text field does not perform its validation untill after blur, as expected. 

Please correct me if I'm wrong but i think this can be closed.

**Comment by theenoahmason** [comment]
**correction:** This problem is only happening on textareas now.

Also incase anyones wondering, currently the latest package as determined by unpkg for example, is 0.28, but the demo for the textfields is currently at 0.29.

**Comment by williamernest** [comment]
Here's a code pen showing the issue for the text field. Setting required via the JS or the HTML5 attribute should behave the same way. 

https://codepen.io/williamernest/pen/ddPdxo

**Comment by theenoahmason** [comment]
Oh I misunderstood, ignore my comments.

**Comment by moog16** [comment]
2 cases that I discovered while playing around with my current fix:

1. mdc-text-field has the invalid class, but required attr is removed. In this instance it seems like the the invalid class on the text field should be removed.
1. mdc-text-field is dirty and has pattern=".{8,}", but not required. Then the required is toggled to true. In this instance I would expect the text field should get the invalid class.

Neither of these two examples are covered in the HTML 5 version. I will speak to designer to see what the correct behavior is.

**Comment by moog16** [comment]
Just spoke with design:

1. Initially if mdc-text-field is dirty and required, then invalid state should be toggled on.
1. Initially if mdc-text-field is not dirty but is required, then do not toggle invalid state.
1. Initially if mdc-text-field is dirty and NOT required, then required is toggled on, then the invalid state should also be toggled on.

**Comment by moog16** [comment]
tl;dr I'm removing `styleValidity_` from setRequired in the foundation.

So after giving it a night to think about this issue, I realized that I have been treating the `setRequired` method the same as a toggle on/off validity method. Meaning if it is required, then all validations should be turned on/off. But there are cases where the input is not required, but if a value exists it should follow validation such as pattern, min/max length (eg. an optional phone number that must match a phone regex). 

Calling `styleValidity_` should not occur in the `setRequired` foundation method. Most likely it should be a public method that allows the end developer to toggle on/off. Otherwise the only other way to trigger validity is to manually add/remove the `.mdc-text-field--invalid` class, which competes with mdc-text-field's foundation.
